# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!


In [40]:
#Import your libraries
import pandas as pd
import numpy as np
import matplotlib as plt

# Challenge 1 -  Explore the Internal Dataset

In this lab, we will start off by working with the wine dataset in scikit-learn. We will select the wine dataset and use a clustering algorithm to learn more about the functionalities of this library. 

We start off by loading the dataset using the `load_wine` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_wine.html)). In the cell below, we will import the function from scikit-learn.

In [41]:
from sklearn.datasets import load_wine

In the cell below, use the `load_wine` function and assign the wine dataset to a variable called `wine`.

In [42]:
# Your code here:
wine=load_wine()

In the next step, list the keys of the variable `wine` to examine its contents. Note that the `load_wine` function does not return dataframes. It returns you a Python dictionary.

In [43]:
# Your code here:
wine.keys()

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])

Next, list the feature names. These are the different characteristics of the wine. 

In [44]:
# Your code here:
wine["feature_names"]

['alcohol',
 'malic_acid',
 'ash',
 'alcalinity_of_ash',
 'magnesium',
 'total_phenols',
 'flavanoids',
 'nonflavanoid_phenols',
 'proanthocyanins',
 'color_intensity',
 'hue',
 'od280/od315_of_diluted_wines',
 'proline']

Print the description of the dataset in the cell below using the `DESCR` attribute of the `wine` variable.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the print function.*

In [45]:
# Your code here:
print(wine["DESCR"])

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

#### From the description, we see that all columns are numeric. We also know that there is no missing data 

Let's plot the alcohol content histogram. Recall that we are working with a numpy array and will need to use a matplotlib function to produce a histogram. 

In [46]:
df_wine=pd.DataFrame.from_dict(wine["data"])

In [47]:
# imports
import plotly.plotly as py
%matplotlib inline
plt.hist(df_wine)

AttributeError: module 'matplotlib' has no attribute 'hist'

In [48]:
# Your code here:


# Challenge 2 - Clustering the Internal Dataset

In this portion of the lab, we will cluster the data to find common traits between the different wines. We will use the k-means clustering algorithm to achieve this goal.

#### We start by importing k-means from scikit-learn and then proceed to create 4 clusters.

In [49]:
# imports
from sklearn.cluster import KMeans

kmeans = KMeans(n_clusters=4)

In [50]:
# Your code here:
census_clusters = kmeans.fit(df_wine)

#### Print the cluster labels.

In [51]:
# Your code here:
a=census_clusters.cluster_centers_

#### Compute the size of each cluster. This can be done by counting the number of occurrences of each unique label in the list above.

Which is the largest cluster of the 4?

In [52]:
# Your code here:
lst=[]
for i in a:
    b=0
    for e in i:
        b=b+1
    lst.append(b)
lst

[13, 13, 13, 13]

In [53]:
# Your answer here:
#Todos los clusters tienen el mismo tamaño

#### Inspect the shape of `wine['data']`

In [54]:
# Your code here:
wine["data"].shape, df_wine.shape

((178, 13), (178, 13))

#### Inspect the first 5 records in `wine['data']`

In [55]:
# Your code here:
df_wine.head()

0     1     2     3      4     5     6     7     8     9     10    11  \
0  14.23  1.71  2.43  15.6  127.0  2.80  3.06  0.28  2.29  5.64  1.04  3.92   
1  13.20  1.78  2.14  11.2  100.0  2.65  2.76  0.26  1.28  4.38  1.05  3.40   
2  13.16  2.36  2.67  18.6  101.0  2.80  3.24  0.30  2.81  5.68  1.03  3.17   
3  14.37  1.95  2.50  16.8  113.0  3.85  3.49  0.24  2.18  7.80  0.86  3.45   
4  13.24  2.59  2.87  21.0  118.0  2.80  2.69  0.39  1.82  4.32  1.04  2.93   

       12  
0  1065.0  
1  1050.0  
2  1185.0  
3  1480.0  
4   735.0

You now know the data object is a 2-dimensional array in which there are 178 rows and 13 columns. Each row is a data record and each column is a feature.

#### What is the average ash content for each cluster? 

*Hints:* 

* *Ash* is the 3rd column.

* The data object is not a Pandas dataframe so you can't apply `pandas.DataFrame.groupby`. Instead, you can use `np.average`.

In [56]:
# Your code here:


# Challenge 3 - Load and Explore an External Dataset

We will now load an external dataset using Pandas and use scikit learn to explore the data. In this portion of the lab, we will use a [patient dataset from Kaggle](https://www.kaggle.com/miles99/patient-admission-dataset-for-learning-data-mining). 

In [121]:
patients = pd.read_csv('../patient-admission-dataset-for-learning-data-mining.csv')

In the next cell, print the first five rows of the data using the `head()` function.

In [122]:
# Your code here:
patients.head()

id         patient_name               patient_email  doctor_phone  \
0   1  Celestyna Dillimore      cdillimore0@dion.ne.jp  674-914-1212   
1   2        Meta Michieli          mmichieli1@loc.gov  172-580-3586   
2   3        Cordie Sancto      csancto2@cafepress.com  794-222-5085   
3   4     Josh De Ambrosis           jde3@amazon.co.jp  856-540-5195   
4   5    Delinda Alfonsini  dalfonsini4@opensource.org  938-978-1131   

  patient_gender patient_dob  patient_diabetic  patient_allergic  \
0         Female  10/18/2018             False              True   
1         Female    2/8/2018             False              True   
2         Female   10/9/2018              True              True   
3           Male   9/10/2018              True              True   
4         Female   2/26/2018             False              True   

   patient_weight_kg  patient_height_sm  patient_nhs_number  \
0                 59                176        8.200152e+09   
1                 77                186                 NaN   
2                 90                177        6.145594e+09   
3                 70                150                 NaN   
4                 82                140        4.804758e+08   

           doctor_name appointment_date  patient_show  is_regular_visit  \
0      Sarena Waliszek         5/1/2018          True              True   
1       Farris Robinet        12/7/2017          True              True   
2       Kaspar Spitaro        10/5/2018         False             False   
3       Rafferty Fowls       10/21/2018         False              True   
4  Glenna MacNeachtain       11/15/2018         False             False   

      prescribed_medicines diagnosis  
0  triamcinolone acetonide      I669  
1                      NaN       NaN  
2                      NaN       NaN  
3                      NaN       NaN  
4                      NaN       NaN

Next, print the column types and check which columns have been misclassified by pandas.

In [123]:
# Your code here:
patients.dtypes

id                        int64
patient_name             object
patient_email            object
doctor_phone             object
patient_gender           object
patient_dob              object
patient_diabetic           bool
patient_allergic           bool
patient_weight_kg         int64
patient_height_sm         int64
patient_nhs_number      float64
doctor_name              object
appointment_date         object
patient_show               bool
is_regular_visit           bool
prescribed_medicines     object
diagnosis                object
dtype: object

#### We can see that none of the date columns have been correctly classified. Also, some columns contain qualitative data that can be dropped.

First, transform the `patient_dob` and `appointment_date` columns to datetime using the `pd.to_datetime` function.

In [124]:
# Your code here:
patients["patient_dob"]=pd.to_datetime(patients["patient_dob"])
patients["appointment_date"]=pd.to_datetime(patients["appointment_date"])

Next, drop the `id`, `patient_name`, `patient_email`, `patient_nhs_number`, and `doctor_phone` columns. These are not quantitative columns and will not contribute to our analysis.

In [125]:
# Your code here:
patients.drop(["id","patient_name","patient_email","patient_nhs_number","doctor_phone"],axis=1,inplace=True)

Now we work on the missing data. Most ML algorithms will not perform as intended if there are missing data.

In the cell below, count how many rows contain missing data in each column. You should see three columns contain missing data:

* `doctor_name`: 58 missing data
* `prescribed_medicines`: 488 missing data
* `diagnosis`: 488 missing data

In [126]:
# Your code here:
patients.isnull().sum()

patient_gender            0
patient_dob               0
patient_diabetic          0
patient_allergic          0
patient_weight_kg         0
patient_height_sm         0
doctor_name              58
appointment_date          0
patient_show              0
is_regular_visit          0
prescribed_medicines    488
diagnosis               488
dtype: int64

The main issues are found in the `prescribed_medicines` and `diagnosis` columns. Can we simply drop these rows?

The answer is not yet. Because when there are missing data in these columns, it doesn't mean the data records are broken. Instead, it means no medication was prescribed and no diagnosis was recorded. Therefore, once we fill in the missing data these columns will be fine. But we'll revisit these columns and decide whether we will eventually drop them when we look at how many unique values are there in these categorical columns.  

For the `prescribed_medicines` column, fill the missing values with the value `no prescription`. For the `diagnosis` column, fill the missing values with `no diagnosis`.

*Hint: Use [`pandas.DataFrame.fillna`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.fillna.html).*

In [127]:
# Your code here:
patients["prescribed_medicines"]=patients["prescribed_medicines"].fillna("no prescription")
patients["diagnosis"]=patients["diagnosis"].fillna("no diagnosis")

How about `doctor_name`? Since a doctor visit without a doctor name might not be meaningful, we will drop these rows.

In [128]:
# Your code here:
patients["doctor_name"].dropna(inplace=True)

#### Another step in preprocessing that can be performed by scikit-learn is label encoding. 

We have 4 columns that are of `bool` type. We would like to convert them to an integer column containing either zero or one. We can do this using [scikit-learn's label encoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelEncoder.html).

In the cell below, import the label encoder and encode the 4 boolean columns (*patient_diabetic*, *patient_allergic*, *patient_show*, *is_regular_visit*) with `0` and `1`. 

In [129]:
# Your code here:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
le.fit(patients["patient_diabetic"])

LabelEncoder()

Print the data dtypes to confirm those four `bool` columns are converted to `int64`.

In [130]:
# Your code here:
patients.dtypes

patient_gender                  object
patient_dob             datetime64[ns]
patient_diabetic                  bool
patient_allergic                  bool
patient_weight_kg                int64
patient_height_sm                int64
doctor_name                     object
appointment_date        datetime64[ns]
patient_show                      bool
is_regular_visit                  bool
prescribed_medicines            object
diagnosis                       object
dtype: object

#### The last step is to handle the `object` data.

There are 4 `object` columns now: `patient_gender`, `doctor_name`, `prescribed_medicines`, and `diagnosis`. The gender columns

In the next cell, check the unique values of each of the `object` columns using `value_counts()`.

In [131]:
# Your code here:
lst=["patient_gender","doctor_name","prescribed_medicines","diagnosis"]
for i in lst:
    print(patients[i].value_counts())

Female    504
Male      496
Name: patient_gender, dtype: int64
Sarena Waliszek           74
Cissy Markey              57
Farris Robinet            55
Flossy Canlin             55
Gabie Brafield            53
Jilly McGrill             53
Glenna MacNeachtain       53
Nerte Elmer               51
Oralie Swaffer            49
Kaspar Spitaro            48
Humfried Cartmel          47
Wyn Grassett              47
Carly SperaJillie Katt    47
Madelina Scurrell         46
Rafferty Fowls            45
Dulce McKerley            42
Rouvin McKinstry          42
Ryley Swallow             40
Hermina Domeny            38
Name: doctor_name, dtype: int64
no prescription                                                                                                                                                                                    488
ALCOHOL                                                                                                                                                    

#### The number of unique values is large for all three columns except `patient_gender`. We will handle these columns differently.

For `diagnosis`, there are too many unique values which will make ML difficult. However, we can re-encode the values to either with or without diagnosis. Remember at an earlier step we filled in the missing values of this column with *no diagnosis*? We can re-encode *no diagnosis* to `0` and all other values to `1`. In this way we can tremendously simply this column.

For `prescribed_medicines`, we can drop this column because it is perfectly correlated with `diagnosis`. Whenever there is no diagnosis, there is no prescribed medicine. So we don't need to keep this duplicated data.

How about `doctor_name`? There are not excessive unique values but still quite many (19). We may either drop or keep it but keeping it will make the analysis more complicated. So due to the length of this lab let's drop it.

How about `gender`? This one is easy. Just like re-encoding the boolean values, we can re-encode gender to `0` and `1` because there are only 2 unique values.

In the next cells, do the following:

1. Create a new column called `diagnosis_int` that has `0` and `1` based on the values in `diagnosis`.

1. Create a new column called `patient_gender_int` that has `0` and `1` based on the values in `patient_gender`.

1. Drop the following columns: `doctor_name`, `diagnosis`, `prescribed_medicines`, and `patient_gender`.

In [132]:
# Your code here:
s = pd.Series(patients["diagnosis"])
s=s.where(s=="no diagnosis", 0)
a=s.where(s!="no diagnosis", 1)
patients["diagnosis_int"]=a.to_frame(name='diagnosis_int')

In [137]:
s = pd.Series(patients["patient_gender"])
s=s.where(s=="Female", 0)
a=s.where(s!="Female", 1)
patients["patient_gender_int"]=a.to_frame(name='patient_gender_int')

In [138]:
patients.drop(["doctor_name","diagnosis","prescribed_medicines","patient_gender"],axis=1, inplace=True)

In [139]:
patients.head()

patient_dob  patient_diabetic  patient_allergic  patient_weight_kg  \
0  2018-10-18             False              True                 59   
1  2018-02-08             False              True                 77   
2  2018-10-09              True              True                 90   
3  2018-09-10              True              True                 70   
4  2018-02-26             False              True                 82   

   patient_height_sm appointment_date  patient_show  is_regular_visit  \
0                176       2018-05-01          True              True   
1                186       2017-12-07          True              True   
2                177       2018-10-05         False             False   
3                150       2018-10-21         False              True   
4                140       2018-11-15         False             False   

  diagnosis_int patient_gender_int  
0             0                  1  
1             1                  1  
2             1                  1  
3             1                  0  
4             1                  1

Let's look at the head again to ensure the re-encoding and dropping are successful:

In [140]:
# Your code here:
patients.dtypes

patient_dob           datetime64[ns]
patient_diabetic                bool
patient_allergic                bool
patient_weight_kg              int64
patient_height_sm              int64
appointment_date      datetime64[ns]
patient_show                    bool
is_regular_visit                bool
diagnosis_int                 object
patient_gender_int            object
dtype: object

An interesting observation is that all patients are no older than 2 years. However, their weights and heights indicate that they are adults. This cannot be true. Therefore, we can either trust the weight and height columns or the DOB column. Since there are other columns that indicate that these are adults (they have emails, some have diabetes) we will drop the `patient_dob` column. We will also drop the `appointment_date` column since it has too many unique values to transform to a dummy variable. Drop the two columns in the cell below.

In [141]:
# Your code here:
patients.drop(["patient_dob","appointment_date"],axis=1, inplace=True)

#### Our data is now ready for clustering. Let's use k-means again.

We start by initializing and fitting a model in the cell below. Call this model patients_cluster.

In [149]:
# Your code here:
kmeans = KMeans(n_clusters=4)
columns=patients.columns
#patients_cluster = kmeans.fit(patients[["patient_weight_kg","patient_height_sm","diagnosis_int","patient_gender_int"]])
patients_cluster = kmeans.fit(patients[columns])
patients_cluster.cluster_centers_

array([[  0.50561798,   0.4906367 ,  94.02996255, 178.92509363,
          0.50187266,   0.53183521,   0.4906367 ,   0.53183521],
       [  0.50617284,   0.49794239,  65.63786008, 152.19341564,
          0.51440329,   0.55555556,   0.48971193,   0.4526749 ],
       [  0.49420849,   0.51737452,  63.1042471 , 176.98841699,
          0.46332046,   0.49420849,   0.54054054,   0.53667954],
       [  0.56277056,   0.51515152,  97.6969697 , 153.76623377,
          0.49350649,   0.51948052,   0.42424242,   0.48917749]])

Attach the labels to the dataframe. Do this by accessing the `labels_` in the `patients_cluster` model and assign them to a new column in `patients` that you will call `labels`.

In [157]:
# Your code here:
labels = patients_cluster.labels_
results = pd.DataFrame([patients.index,labels]).T


Now using a `groupby`, find the mean of every variable in `patients` and group by the `labels` column. This summary will allow us to see how the patients differ between the clusters. Your output should look similar to the image below.

![groupby mean](../groupby-mean.png)

Additionally, add a comment to describe which columns have the largest difference between clusters.

In [ ]:
# Your code here:


In [ ]:
# Your comment here:


# Bonus Challenge: Visualize K-Means Clusters

How did k-means cluster the data? You can obtain an intuitive view with a scatter plot. Generate a 2-d cluster plot below using `matplotlib`. You need to choose 2 of the features from your cleaned and transformed dataset, and use color to represent the cluster label generated from k-means.

If the scatter plot does not make any sense to you, it means the features you chose to visualize are not the right ones. You should be able to see 4 clear clusters with different colors in your visualization that suggests how k-means had clustered your data.

![Cluster Visualization](../clusters.png)

In [ ]:
# Your code here:


Additionally, you can visualize the clusters in 3-D scatter plot. Give it a try below.

In [ ]:
# Your code here:
